cfy_manager/components/cluster.py

```diff
@@ -12,25 +12,159 @@
+        if response.code == 401:
+            raise NetworkError(
+                'Could not connect to the REST service: '
+                '401 unauthorized. Possible access control misconfiguration'
+            )
+        if response.code != 200:
+            raise NetworkError(
+                'REST service returned an unexpected response: '
+                '{0}'.format(response.code)
+            )
+
+        try:
+            return json.load(response)
+        except ValueError as e:
+            raise BootstrapError(
+                'REST service returned malformed JSON: {0}'.format(e))
+
+    def _verify_premium(self):
+        try:
+            find_module('cloudify_premium')
+        except ImportError as e:
+            premium_enabled = False
+        else:
+            premium_enabled = True
+        return premium_enabled
+
+    def _get_join_addresses(self, master_manager_ip):
+        result = self._generic_cloudify_rest_request(
+            master_manager_ip,
+            config[RESTSERVICE]['port'],
+            'v3.1/cluster/nodes',
+            'get')
+        return [n.host_ip for n in result['items']]
+
+    def _adding_new_cluster_node_on_master(self, master_manager_ip,
+                                           slave_host_ip):
+        data = {
+            'host_ip': master_manager_ip,
+            'node_name': slave_host_ip
+        }
+        result = self._generic_cloudify_rest_request(
+            master_manager_ip,
+            config[RESTSERVICE]['port'],
+            'v3.1/cluster/nodes/{0}'.format(slave_host_ip),
+            'put',
+            data
+        )
+        return result
+
+    def _get_cluster_status(self, master_manager_ip):
+        result = self._generic_cloudify_rest_request(
+            master_manager_ip,
+            config[RESTSERVICE]['port'],
+            'v3.1/cluster',
+            'get')
+        return result
+
+    def _join_to_cluster(self, master_manager_ip):
+        cluster_node_host_ip = config[MANAGER][PUBLIC_IP]
+        cluster_node_name = config[CLUSTER][NODE_NAME]
+        join_addrs = self._get_join_addresses(master_manager_ip)
+        new_node = self._adding_new_cluster_node_on_master(master_manager_ip,
+                                                           join_addrs)
+        data = {
+            'host_ip': cluster_node_host_ip,
+            'node_name': cluster_node_name,
+            'credentials': new_node['credentials'],
+            'join_addrs': join_addrs,
+            'ignore_database_validation': False,
+            'bootstrap_cluster': False
+        }
+
+        start_cluster = os.path.join(
+            os.path.dirname(sys.executable), 'create_cluster_node')
+
+        with tempfile.NamedTemporaryFile(delete=False) as f:
+            json.dump(data, f)
+
+        # Same as a regular all-in-one installation, we run the sub-process
+        # instead of a rest call
+        create_cluster_node = ['systemd-run', 'cloudify-ha-cluster',
```

Comment:
this doesn't (shouldn't) require systemd-run
systemd-run is used in restservice so that it's run 1) as root, 2) not as a child process of the restservice, 3) is non-blocking for the request
here we should be able to just `common.sudo([start_cluster])`

Reply:
fixed